File: wpimath/src/main/native/cpp/system/struct/GearboxStruct.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/system/struct/GearboxStruct.hpp"

namespace {
constexpr size_t kMotorOff = 0;
constexpr size_t kNumMotorsOff =
    kMotorOff + wpi::util::GetStructSize<wpi::math::DCMotor>();
constexpr size_t kReductionOff = kNumMotorsOff + 4;
}  // namespace

using StructType = wpi::util::Struct<wpi::math::Gearbox>;

wpi::math::Gearbox StructType::Unpack(std::span<const uint8_t> data) {
  return wpi::math::Gearbox{
      wpi::util::UnpackStruct<wpi::math::DCMotor, kMotorOff>(data),
      wpi::util::UnpackStruct<int32_t, kNumMotorsOff>(data),
      wpi::util::UnpackStruct<double, kReductionOff>(data),
  };
}

void StructType::Pack(std::span<uint8_t> data,
                      const wpi::math::Gearbox& value) {
  wpi::util::PackStruct<kMotorOff>(data, value.motor);
  wpi::util::PackStruct<kNumMotorsOff>(data,
                                       static_cast<int32_t>(value.numMotors));
  wpi::util::PackStruct<kReductionOff>(data, value.reduction);
}

File: wpimath/src/main/java/org/wpilib/math/system/struct/GearboxStruct.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.system.struct;

import java.nio.ByteBuffer;
import org.wpilib.math.system.DCMotor;
import org.wpilib.math.system.Gearbox;
import org.wpilib.util.struct.Struct;

public class GearboxStruct implements Struct<Gearbox> {
  @Override
  public Class<Gearbox> getTypeClass() {
    return Gearbox.class;
  }

  @Override
  public String getTypeName() {
    return "Gearbox";
  }

  @Override
  public int getSize() {
    return DCMotor.struct.getSize() + INT32_SIZE + DOUBLE_SIZE;
  }

  @Override
  public String getSchema() {
    return "DCMotor motor;int32 num_motors;double reduction";
  }

  @Override
  public Struct<?>[] getNested() {
    return new Struct<?>[] {DCMotor.struct};
  }

  @Override
  public Gearbox unpack(ByteBuffer bb) {
    DCMotor motor = DCMotor.struct.unpack(bb);
    int numMotors = bb.getInt();
    double reduction = bb.getDouble();
    return new Gearbox(motor, numMotors, reduction);
  }

  @Override
  public void pack(ByteBuffer bb, Gearbox value) {
    DCMotor.struct.pack(bb, value.motor);
    bb.putInt(value.numMotors);
    bb.putDouble(value.reduction);
  }

  @Override
  public boolean isImmutable() {
    return true;
  }
}
